lib/ansible/modules/cloud/vmware/vmware_guest_register_operation.py

```diff
@@ -0,0 +1,277 @@
+    name: "{{ vm_name }}"
+    template: no
+    path: "[datastore1] vm/vm.vmx"
+    state: present
+
+- name: UnRegister VM from inventory
+  vmware_guest_register_operation:
+    hostname: "{{ vcenter_hostname }}"
+    username: "{{ vcenter_username }}"
+    password: "{{ vcenter_password }}"
+    validate_certs: no
+    datacenter: "{{ datacenter }}"
+    folder: "/vm"
+    name: "{{ vm_name }}"
+    state: absent
+'''
+
+RETURN = r'''
+'''
+
+try:
+    from pyVmomi import vim, vmodl
+    HAS_PYVMOMI = True
+except ImportError:
+    HAS_PYVMOMI = False
+
+from ansible.module_utils._text import to_native
+from ansible.module_utils.vmware import PyVmomi, vmware_argument_spec, find_resource_pool_by_name, \
+    wait_for_task, compile_folder_path_for_object, find_cluster_by_name
+from ansible.module_utils.basic import AnsibleModule
+
+
+class VMwareGuestRegisterOperation(PyVmomi):
+    def __init__(self, module):
+        super(VMwareGuestRegisterOperation, self).__init__(module)
+        self.datacenter = module.params["datacenter"]
+        self.cluster = module.params["cluster"]
+        self.folder = module.params["folder"]
+        self.name = module.params["name"]
+        self.esxi_hostname = module.params["esxi_hostname"]
+        self.path = module.params["path"]
+        self.template = module.params["template"]
+        self.resource_pool = module.params["resource_pool"]
+        self.state = module.params["state"]
+
+    def execute(self):
+        result = dict(changed=False)
+
+        datacenter = self.find_datacenter_by_name(self.datacenter)
+        if not datacenter:
+            self.module.fail_json(msg="Cannot find the specified Datacenter: %s" % self.datacenter)
+
+        dcpath = compile_folder_path_for_object(datacenter)
+        if not dcpath.endswith("/"):
+            dcpath += "/"
+
+        if(self.folder in [None, "", "/"]):
+            self.module.fail_json(msg="Please specify folder path other than blank or '/'")
+        elif(self.folder.startswith("/vm")):
+            fullpath = "%s%s%s" % (dcpath, self.datacenter, self.folder)
+        else:
+            fullpath = "%s%s" % (dcpath, self.folder)
+
+        folder_obj = self.content.searchIndex.FindByInventoryPath(inventoryPath="%s" % fullpath)
+        if not folder_obj:
+            details = {
+                'datacenter': datacenter.name,
+                'datacenter_path': dcpath,
+                'folder': self.folder,
+                'full_search_path': fullpath,
+            }
+            self.module.fail_json(msg="No folder %s matched in the search path : %s" % (self.folder, fullpath),
+                                  details=details)
+
+        if self.state == "present":
+            if self.get_vm():
+                self.module.exit_json(**result)
+            else:
+                if self.esxi_hostname:
+                    host_obj = self.find_hostsystem_by_name(self.esxi_hostname)
+                    if not host_obj:
```

Comment:
This is the 4th `if` level, it probably means to can re-factorize your code a bit, e.g: with a function, of smaller if/else block.

This being said, the code is short enough to be comprehensible, I don't mind merging it like it is today.